lib/portage/sync/modules/zipfile/zipfile.py

```diff
@@ -0,0 +1,104 @@
+import logging
+import zipfile
+import shutil
+import tempfile
+import urllib.request
+
+import portage
+from portage.util import writemsg_level, writemsg_stdout
+from portage.sync.syncbase import SyncBase
+
+
+class ZipFile(SyncBase):
+    """ZipFile sync module"""
+
+    short_desc = "Perform sync operations on GitHub repositories"
+
+    @staticmethod
+    def name():
+        return "ZipFile"
+
+    def __init__(self):
+        SyncBase.__init__(self, "emerge", ">=sys-apps/portage-2.3")
+
+    def sync(self, **kwargs):
+        """Sync the repository"""
+        if kwargs:
+            self._kwargs(kwargs)
+
+        req = urllib.request.Request(url=self.repo.sync_uri)
+
+        info = portage.grabdict(os.path.join(self.repo.location, ".info"))
+        if "etag" in info:
+            req.add_header("If-None-Match", info["etag"][0])
+
+        try:
+            with urllib.request.urlopen(req) as response:
+                with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
+                    shutil.copyfileobj(response, tmp_file)
+
+                zip_file = tmp_file.name
+                etag = response.headers.get("etag")
+
+        except urllib.error.HTTPError as resp:
+            if resp.code == 304:
+                writemsg_stdout(">>> The repository has not changed.\n", noiselevel=-1)
+                return (os.EX_OK, False)
+
+            writemsg_level(
+                f"!!! Unable to obtain zip archive: {resp}\n",
+                noiselevel=-1,
+                level=logging.ERROR,
+            )
+            return (1, False)
+
+        if not zipfile.is_zipfile(zip_file):
+            msg = "!!! file is not a zip archive."
+            self.logger(self.xterm_titles, msg)
+            writemsg_level(msg + "\n", noiselevel=-1, level=logging.ERROR)
+
+            os.unlink(zip_file)
+
+            return (1, False)
+
+        # Drop previous tree
+        shutil.rmtree(self.repo.location)
+
+        with zipfile.ZipFile(zip_file) as archive:
+            strip_comp = 0
+
+            for f in archive.namelist():
+                f = os.path.normpath(f)
+                if os.path.basename(f) == "profiles":
+                    strip_comp = f.count("/")
+                    break
+
+            for n in archive.infolist():
+                p = os.path.normpath(n.filename)
+
+                if os.path.isabs(p):
+                    continue
+
+                parts = p.split("/")
+                dstpath = os.path.join(self.repo.location, *parts[strip_comp:])
+
+                if n.is_dir():
+                    os.makedirs(dstpath, mode=0o755, exist_ok=True)
+                    continue
+
+                with archive.open(n) as srcfile:
+                    with open(dstpath, "wb") as dstfile:
+                        shutil.copyfileobj(srcfile, dstfile)
+
+        with open(os.path.join(self.repo.location, ".info"), "w") as infofile:
+            if etag:
+                infofile.write(f"etag {etag}\n")
```

Comment:
We can implement the retrieve_head method to retrieve the etag like this:
```python
    def retrieve_head(self, **kwargs):
        """Get information about the head commit"""
        if kwargs:
            self._kwargs(kwargs)
        info = portage.grabdict(os.path.join(self.repo.location, ".info"))
        if "etag" in info:
            return (os.EX_OK, info["etag"][0])
        return (1, False)
```

Reply:
@zmedico Good catch! I'll add this as a separate commit.

Reply:
I went ahead and wrote a little test case that calls `emerge --sync` and checks the etag afterwards. It would be nice to include in the PR so we have some test coverage from the start.

Signed-off-by: Zac Medico <zmedico@gentoo.org>
```python
# Copyright 2024 Gentoo Authors

import http.server
import os
import shutil
import socketserver
import subprocess
import tempfile
import textwrap
import threading
from functools import partial

import portage
from portage.tests import TestCase
from portage.tests.resolver.ResolverPlayground import ResolverPlayground


class test_sync_zipfile_case(TestCase):
    def test_sync_zipfile(self):
        cpv = "dev-libs/A-0"
        ebuilds = {
            cpv: {"EAPI": "8"},
        }
        etag = "foo"

        server = None
        playground = None
        tmpdir = tempfile.mkdtemp()
        try:

            class Handler(http.server.SimpleHTTPRequestHandler):
                def end_headers(self):
                    self.send_header("etag", etag)
                    super().end_headers()

            server = socketserver.TCPServer(
                ("127.0.0.1", 0),
                partial(Handler, directory=tmpdir),
            )
            threading.Thread(target=server.serve_forever, daemon=True).start()

            playground = ResolverPlayground(
                ebuilds=ebuilds,
            )
            settings = playground.settings

            env = settings.environ()

            repos_conf = textwrap.dedent(
                """
                [test_repo]
                location = %(location)s
                sync-type = zipfile
                sync-uri = %(sync-uri)s
                auto-sync = true
            """
            )

            repo_location = f"{playground.eprefix}/var/repositories/test_repo"

            env["PORTAGE_REPOSITORIES"] = repos_conf % {
                "location": repo_location,
                "sync-uri": "http://{}:{}/test_repo.zip".format(*server.server_address),
            }

            shutil.make_archive(os.path.join(tmpdir, "test_repo"), "zip", repo_location)

            ebuild = playground.trees[playground.eroot]["porttree"].dbapi.findname(cpv)
            self.assertTrue(os.path.exists(ebuild))
            shutil.rmtree(repo_location)
            self.assertFalse(os.path.exists(ebuild))

            result = subprocess.run(
                [
                    "emerge",
                    "--sync",
                ],
                env=env,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
            )
            output = result.stdout.decode(errors="replace")
            try:
                self.assertEqual(result.returncode, os.EX_OK)
            except Exception:
                print(output)
                raise

            repo = settings.repositories["test_repo"]
            sync_mod = portage.sync.module_controller.get_class("zipfile")
            status, repo_revision = sync_mod().retrieve_head(options={"repo": repo})
            self.assertEqual(status, os.EX_OK)
            self.assertEqual(repo_revision, etag)
        finally:
            if server is not None:
                server.shutdown()
            shutil.rmtree(tmpdir)
            playground.cleanup()
```

Reply:
@zmedico I'm not very familiar with the portage codebase, but I added your test as a separate commit.

Reply:
Great, thanks!